src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ResettableValueTaskSource.cs

```diff
@@ -91,11 +92,11 @@ public bool TryGetValueTask(out ValueTask valueTask, object? keepAlive = null, C
                     _cancellationRegistration = cancellationToken.UnsafeRegister(static (obj, cancellationToken) =>
                     {
                         (ResettableValueTaskSource thisRef, object? target) = ((ResettableValueTaskSource, object?))obj!;
-                        // This will transition the state to Ready.
-                        if (thisRef.TrySetException(new OperationCanceledException(cancellationToken)))
+                        lock (thisRef)
                         {
-                            thisRef._cancellationAction?.Invoke(target);
+                            thisRef._cancelledToken = cancellationToken;
```

Comment:
We don't set the task from cancellation, now it's part of cancellation action. But we remember that it was cancelled via this token and throw appropriate error in GetResult, effectively overriding any result set to the task.
This helps with waiting on `SEND_COMPLETE` in `WriteAsync` where we don't want cancellation to unblock the task, but we still want to propagate it to the user when `SEND_COMPLETE` arrives.